src/ILCompiler.ReadyToRun/src/Compiler/ReadyToRunMetadataFieldLayoutAlgorithm.cs

```diff
@@ -227,6 +175,251 @@ protected override ModuleFieldLayout CreateValueFromKey(EcmaModule module)
+                    case CorElementType.ELEMENT_TYPE_STRING:
+                    case CorElementType.ELEMENT_TYPE_SZARRAY:
+                    case CorElementType.ELEMENT_TYPE_ARRAY:
+                    case CorElementType.ELEMENT_TYPE_CLASS:
+                    case CorElementType.ELEMENT_TYPE_OBJECT:
+                        isGcPointerField = true;
+                        break;
+
+                    case CorElementType.ELEMENT_TYPE_BYREF:
+                        ThrowHelper.ThrowTypeLoadException(ExceptionStringID.ClassLoadGeneral, fieldDesc.OwningType);
+                        break;
+
+                    // Statics for valuetypes where the valuetype is defined in this module are handled here. 
+                    // Other valuetype statics utilize the pessimistic model below.
+                    case CorElementType.ELEMENT_TYPE_VALUETYPE:
+                        isGcBoxedField = true;
+                        if (IsTypeByRefLike(valueTypeHandle, module.MetadataReader))
+                        {
+                            ThrowHelper.ThrowTypeLoadException(ExceptionStringID.ClassLoadGeneral, fieldDesc.OwningType);
+                        }
+                        break;
+
+                    case CorElementType.ELEMENT_TYPE_END:
+                    default:
+                        isGcBoxedField = true;
+                        if (!valueTypeHandle.IsNil)
+                        {
+                            // Allocate pessimistic non-GC area for cross-module fields as that's what CoreCLR does
+                            // <a href="https://github.com/dotnet/coreclr/blob/659af58047a949ed50d11101708538d2e87f2568/src/vm/ceeload.cpp#L2124">here</a>
+                            alignment = TargetDetails.MaximumPrimitiveSize;
+                            size = TargetDetails.MaximumPrimitiveSize;
+                        }
+                        else
+                        {
+                            // Field has an unexpected type
+                            throw new InvalidProgramException();
+                        }
+                        break;
+                }
+            }
+
+            public FieldAndOffset[] GetOrAddDynamicLayout(DefType defType, ModuleFieldLayout moduleFieldLayout)
+            {
+                FieldAndOffset[] fieldsForType;
+                if (!moduleFieldLayout.TryGetDynamicLayout(defType, out fieldsForType))
+                {
+                    fieldsForType = CreateDynamicLayout(defType, moduleFieldLayout.Module);
+                    moduleFieldLayout.AddDynamicLayout(defType, fieldsForType);
+                }
+                return fieldsForType;
+            }
+
+            private FieldAndOffset[] CreateDynamicLayout(DefType defType, EcmaModule module)
+            {
+                List<FieldAndOffset> fieldsForType = null;
+                int pointerSize = module.Context.Target.PointerSize;
+
+                int[][] nonGcStaticsCount = new int[][]
```

Comment:
One other post-merge comment, I'm not sure how much this dynamic layout algorithm will change, but it would helpful to clarify that the first index of your jagged array structures determines if you're using non-ThreadStatic and ThreadStatic fields.